### PR TITLE
Virtual Private Cloud with subnetwork [BA-5676]

### PR DESCRIPTION
**Pull request by salonishah11**

### CODE REVIEW

File: centaur/src/main/resources/standardTestCases/virtual_private_cloud/check_network_in_vpc.wdl [hunk 1]
```diff
@@ -2,12 +2,17 @@ version 1.0
...
   command {
+    set -euo pipefail
+
     apt-get install --assume-yes jq > /dev/null
     INSTANCE=$(curl -s "http://metadata.google.internal/computeMetadata/v1/instance/name" -H "Metadata-Flavor: Google")
     ZONE=$(curl -s "http://metadata.google.internal/computeMetadata/v1/instance/zone" -H "Metadata-Flavor: Google" | sed -E 's!.*/(.*)!\1!')
     TOKEN=$(gcloud auth application-default print-access-token)
     INSTANCE_METADATA=$(curl "https://www.googleapis.com/compute/v1/projects/broad-dsde-cromwell-dev/zones/$ZONE/instances/$INSTANCE" -H "Authorization: Bearer $TOKEN" -H 'Accept: application/json')
-    echo $INSTANCE_METADATA | jq -r '.networkInterfaces[0].network' | sed -E 's!.*/(.*)!\1!'
+    NETWORK_OBJECT=`echo $INSTANCE_METADATA | jq -r '.networkInterfaces[0]'`
```
Comment (kshakir) [inline centaur/src/main/resources/standardTestCases/virtual_private_cloud/check_network_in_vpc.wdl]:
```shell
NETWORK_OBJECT=$(echo $INSTANCE_METADATA | jq  --exit-status --raw-output '.networkInterfaces[0]')
```

Suggestions:
- Use `$()` instead of backticks, as the former may be [nested](https://stackoverflow.com/questions/9449778/what-is-the-benefit-of-using-instead-of-backticks-in-shell-scripts#answer-9449788)
- Get in the habit of adding `--exit-status` to `jq` to stop it from [silently exiting without error](https://stedolan.github.io/jq/manual/v1.6/#Invokingjq)
- (Debatable) Short args like `-r` are meant for [typing quickly on the command line](https://superuser.com/questions/174564/why-are-there-short-and-long-alternatives-for-command-line-options)

Same goes for the other `jq` calls on the lines below.

Bonus ToL: If `echo ... | jq ...` is working, leave it. But in general I don't know if I [trust `echo` not to modify the contents of variables](https://stackoverflow.com/questions/40027395/passing-bash-variable-to-jq-select#comment67406349_40027637). I don't have the exact change ready at the moment, but there are discussions around the web of [other ways of inputing values into `jq`](https://stackoverflow.com/questions/47105490/can-i-pass-a-string-variable-to-jq-not-the-file#answer-47106137).

Comment (salonishah11) [reply centaur/src/main/resources/standardTestCases/virtual_private_cloud/check_network_in_vpc.wdl]: Modified as suggested. Thank you!

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributes.scala [hunk 2]
```diff
@@ -154,11 +157,12 @@ object PipelinesApiConfigurationAttributes {
...
 
     val vpcNetworkLabel: ErrorOr[Option[String]] = validate { backendConfig.getAs[String]("virtual-private-cloud.network-label-key") }
+    val vpcSubnetworkLabel: ErrorOr[Option[String]] = validate { backendConfig.getAs[String]("virtual-private-cloud.subnetwork-label-key") }
     val vpcAuth: ErrorOr[Option[String]] = validate { backendConfig.getAs[String]("virtual-private-cloud.auth")}
 
     val virtualPrivateCloudConfiguration: ErrorOr[Option[VirtualPrivateCloudConfiguration]] = {
-      (vpcNetworkLabel, vpcAuth) flatMapN  {
-        case (network, auth) => validateVPCConfig(network, auth)
+      (vpcNetworkLabel, vpcSubnetworkLabel, vpcAuth) flatMapN  {
+        case (network, subnetwork, auth) => validateVPCConfig(network, subnetwork, auth)
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributes.scala]: you _might_ be able to say `... flatMapN validateVPCConfig` here

File: supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala [hunk 3]
```diff
@@ -88,15 +88,22 @@ case class GenomicsFactory(applicationName: String, authMode: GoogleAuthMode, en
...
+          case (Some(networkLabel), Some(subnetworkLabel)) =>
             new Network()
               .setUsePrivateAddress(createPipelineParameters.runtimeAttributes.noAddress)
               .setName(VirtualPrivateCloudNetworkPath.format(createPipelineParameters.projectId, networkLabel._2))
-          case None =>
-            jobLogger.debug("Project `{}` does not have high security network configured. " +
-              "Project metadata does not have network label key `{}`. Falling back to running the job on default network.", createPipelineParameters.projectId: Any, vpcConfig.name: Any)
+              .setSubnetwork(subnetworkLabel._2)
+
+          case (Some(networkLabel), None) =>
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala]: There's a fair amount of repetition between these cases. I think you could simplify with something like:
```scala
          case (Some(networkLabel), subnetworkLabel) =>
            val network = new Network()
              .setUsePrivateAddress(createPipelineParameters.runtimeAttributes.noAddress)
              .setName(VirtualPrivateCloudNetworkPath.format(createPipelineParameters.projectId, networkLabel._2))
            
            subnetworkLabel foreach { case (_, subnet) => network.setSubnetwork(subnet) }
            network
```

Comment (salonishah11) [reply supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala]: Is it possible to update `val network` object's property once created?

Comment (cjllanwarne) [reply supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala]: Clicking through to the implementation of the function, it looks like it's just a normal mutable java class behind the scenes

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributes.scala (hunk 2 above)
Comment (salonishah11) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributes.scala]: TIL. Thank you!

File: supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala (hunk 3 above)
Comment (salonishah11) [reply supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala]: Ahh I see. That's right. I have removed the repetition. Thank you.

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributes.scala [hunk 4]
```diff
@@ -202,7 +204,8 @@ object PipelinesApiConfigurationAttributes {
     }
 
     (project, executionBucket, endpointUrl, genomicsAuthName, genomicsRestrictMetadataAccess, gcsFilesystemAuthName,
-      qpsValidation, duplicationStrategy, requestWorkers, localizationConfiguration, virtualPrivateCloudConfiguration, batchRequestTimeoutConfigurationValidation) flatMapN authGoogleConfigForPapiConfigurationAttributes match {
+      qpsValidation, duplicationStrategy, requestWorkers, localizationConfiguration, virtualPrivateCloudConfiguration,
+      batchRequestTimeoutConfigurationValidation) flatMapN authGoogleConfigForPapiConfigurationAttributes match {
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributes.scala]: TOL (my personal preference) if you split one of these onto a new line, I prefer if we split the entire list onto newlines

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributes.scala [hunk 5]
```diff
@@ -101,16 +102,18 @@ object PipelinesApiConfigurationAttributes {
...
+    def validateVPCConfig(networkOption: Option[String], subnetworkOption: Option[String], authOption: Option[String]): ErrorOr[Option[VirtualPrivateCloudConfiguration]] = {
+      (networkOption,subnetworkOption, authOption) match {
+        case (Some(network), _, Some(auth)) => googleConfig.auth(auth) match {
+          case Valid(validAuth) => Option(VirtualPrivateCloudConfiguration(network, subnetworkOption, validAuth)).validNel
+          case Invalid(error) => s"Auth $auth is not valid for Virtual Private Cloud configuration. Reason: $error" .invalidNel
+        }
+        case (Some(_), _, None) => vpcErrorMessage(List("auth"))
+        case (None, _, Some(_)) => vpcErrorMessage(List("network-label-key"))
+        case (None, Some(_), None) => vpcErrorMessage(List("network-label-key", "auth"))
+        case (None, None, None) => None.validNel
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributes.scala]: I see you've added tests for `Some`/`Some`/`Some` and `None`/`Some`/`None`. Do we need tests for any of the other code paths (eg `Some`/`None`/`Some`?)

File: supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala [hunk 6]
```diff
@@ -88,15 +88,18 @@ case class GenomicsFactory(applicationName: String, authMode: GoogleAuthMode, en
 
       def networkFromLabels(vpcConfig: VirtualPrivateCloudConfiguration, projectLabels: ProjectLabels): Network = {
         val networkLabelOption = projectLabels.labels.find(l => l._1.equals(vpcConfig.name))
+        val subnetworkLabelOption = vpcConfig.subnetwork.flatMap(s => projectLabels.labels.find(l => l._1.equals(s)))
 
-        networkLabelOption match {
-          case Some(networkLabel) =>
-            new Network()
+        (networkLabelOption, subnetworkLabelOption) match {
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala]: No need to add `subnetworkLabelOption` to the match if you just use `_` in both `case`s

File: supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala [hunk 7]
```diff
@@ -88,15 +88,18 @@ case class GenomicsFactory(applicationName: String, authMode: GoogleAuthMode, en
...
               .setUsePrivateAddress(createPipelineParameters.runtimeAttributes.noAddress)
               .setName(VirtualPrivateCloudNetworkPath.format(createPipelineParameters.projectId, networkLabel._2))
-          case None =>
-            jobLogger.debug("Project `{}` does not have high security network configured. " +
-              "Project metadata does not have network label key `{}`. Falling back to running the job on default network.", createPipelineParameters.projectId: Any, vpcConfig.name: Any)
+
+            subnetworkLabelOption foreach { case(_, subnet) => network.setSubnetwork(subnet) }
+            network
+          case (None, _) =>
+            // Falling back to running the job on default network since the project does not have high security network configured
```
Comment (cjllanwarne) [inline supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala]: TOL: Technically, ... `"... since the project does not provide the network specifying keys in its metadata"`.

In one project the network specified may not be high security, and in another the default network might be high security.

File: CHANGELOG.md [hunk 8]
```diff
@@ -2,6 +2,11 @@
 
 ## 43 Release Notes
 
+### Virtual Private Cloud with Subnetworks
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply CHANGELOG.md]: Hmm so does this change get mentioned twice in 2 separate sections? Once in `Configuration changes` and other as it's own heading?

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributes.scala (hunk 4 above)
Comment (salonishah11) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributes.scala]: I have split the entire list by newlines 👍

File: supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributes.scala (hunk 5 above)
Comment (salonishah11) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributes.scala]: The `(Some, None, Some)` already has a test from the original VPC with only network label key PR. (https://github.com/broadinstitute/cromwell/blob/412fea17b028787e4cd9051e2ef7ddc49e393805/supportedBackends/google/pipelines/common/src/test/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributesSpec.scala#L121-L136)

Comment (salonishah11) [reply supportedBackends/google/pipelines/common/src/main/scala/cromwell/backend/google/pipelines/common/PipelinesApiConfigurationAttributes.scala]: I will add test for the remaining ones 👍

File: supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala (hunk 6 above)
Comment (salonishah11) [reply supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala]: Hmm that's right. I guess I missed that while refactoring it. I have removed it. Thank you.

File: supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala (hunk 7 above)
Comment (salonishah11) [reply supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala]: >In one project the network specified may not be high security, and in another the default network might be high security.

I am not sure if I am completely following this. So are you saying that someone's `network-label-key` might point to a default network rather than a high security one?

File: centaur/src/main/resources/standardTestCases/check_network_in_vpc.test [hunk 9]
```diff
@@ -4,11 +4,14 @@ backends: [Papiv2-Virtual-Private-Cloud]
...
   workflow: virtual_private_cloud/check_network_in_vpc.wdl
+  options: virtual_private_cloud/wf_zone_options.json
 }
 
 metadata {
   workflowName: check_network_in_vpc
   status: Succeeded
 
   "outputs.check_network_in_vpc.network_used": "cromwell-ci-vpc-network"
+  "outputs.check_network_in_vpc.subnetwork_used": "cromwell-ci-vpc-network"
```
[the remark this answers is not in the document]
Comment (salonishah11) [reply centaur/src/main/resources/standardTestCases/check_network_in_vpc.test]: >Is this right? The subnetwork is called the same as the network?

Yes. While creating a custom network, when you allow the service to automatically create subnets in all the regions, the name of the custom network and subnetwork is the same. I had problems while creating a custom network and manually creating subnets and their associated IPs in all the regions. So I switched back to the auto mode for subnet creations. That's why I added an additional check for the zone to make sure that the task runs on a particular subnet in a particular region.

File: supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala (hunk 7 above)
Comment (cjllanwarne) [reply supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala]: I'm just saying that "custom network" does not imply high security, and also that "default network" does not necessarily imply low security

Comment (salonishah11) [reply supportedBackends/google/pipelines/v2alpha1/src/main/scala/cromwell/backend/google/pipelines/v2alpha1/GenomicsFactory.scala]: Ah I see. I will edit that comment.

**Comment by salonishah11** [comment]
>do you have a test for the network and auth specified, but no subnetwork?

@cjllanwarne do you mean a centaur test? Or unit test?

**Comment by salonishah11** [comment]
>do you have a test for the network and auth specified, but no subnetwork

Talked offline. The unit test for this is already added.
